Step3 - 로또(자동) 리뷰 요청드립니다.

programmer-sjk opened the pull request:
안녕하세요 민수님. 
피드백 반영한 로또(자동) 리뷰 부탁드립니다.

### domain 설명
`Rank`: 순위 별 상금 정보
`LottoResult`: 당첨 티켓 및 통계와 수익률 담당
`LottoMarket`: 로또를 판매하는 가게
`LottoNumber`: 로또 하나의 번호
`LottoTicket`: 6개의 로또 번호를 가진 티켓
`Money`: 원시값 포장한 객체

### 조금 고민했던 사항
* Controller에 `int money = ~~` 와 같은 코드가 나와 원시값 포장해야지. 하다가 모든 로직의 숫자를 Money 객체로 만들었는데, commit 후 돌아가기 힘든 강을 건넌 느낌입니다. 과연 좋은 방법일지 민수님은 어떻게 생각하실까요.
* enum 패키지는 예약어라 enums 패키지를 만들었는데 보통 이렇게 사용하는지 궁금하네요.
* 전체 상금을 계산하는 주체가, Rank 클래스가 상금 정보를 가지고 있어 Rank 클래스가 `calculatePrice` 메소드를 가지고 있는데 괜찮았을까요..?

### 요청사항
* for 문이 Stream 보다 빠르니까 이해하기 쉬운 코드들은 for 문을 유지했는데 혹시 더 나이스하게 풀 수 있다면 꼭 리뷰를 남겨주시면 숙지 하겠습니다.

programmer-sjk committed:
refactor: 2단계 피드백 반영

코드 리뷰 내용 반영
- 매직 넘버 상수화
- String.split 상수로 선언
- early return 적용
- reduce -> sum 방식으로 적용
- NullAndEmptySource 어노테이션 사용
- 변수 result 대신 actual 사용

programmer-sjk committed:
docs(README.md): 요구사항 및 기능 구현 목록 작성

programmer-sjk committed:
feat: 구입 금액을 입력받는다.

프로그램 구조 및 금액을 입력받는 기능 개발
- 메인 파일(Application) 추가
- Controller 추가
- UI를 담당하는 InputView, ResultView 추가

programmer-sjk committed:
feat: 금액만큼 로또를 자동생성한다

전반적인 도메인 추가
- 로또 번호, 로또 티켓, 로또 마켓 도메인 추가
- Lotto 티켓을 생성하는 Generator 추가
- 테스트 코드 작성

programmer-sjk committed:
feat: 생성한 로또 번호를 출력한다

programmer-sjk committed:
feat: 당첨 번호를 입력받는다

programmer-sjk committed:
feat: 당첨 번호와 숫자가 몇 개 일치하는지 통계를 구한다.

통계 구하는 기능 개발에 따른 도메인 및 테스트 코드 추가
- LottoCommittee (협회) 도메인 추가
- Rank enum 추가
- 통계 Dto 추가
- 수동으로 로또를 생성하는 Generator 추가
- test 코드 및 fixture 추가

programmer-sjk committed:
feat: 수익률을 구해서 출력한다

수익률 기능에 따른 기능 추가 및 테스트 코드 추가
- LottoCommittee 수익률 구하는 메소드 추가
- Rank enum에 전체 가격을 반환하는 함수 추가
- 테스트 코드 작성

programmer-sjk committed:
refactor(domain): dto, enums 패키지 추가

programmer-sjk committed:
refactor: 입력받은 돈을 원시값으로 포장

Money 객체로 입력받은 돈을 포장한다.
- Money 객체 추가
- Rank에서 관련된 상금을 Money 객체로 관리
- 테스트 코드 수정

programmer-sjk committed:
refactor: 사용하지 않는 import 문 제거

programmer-sjk committed:
refactor: ManualLottoGenerator 사용

당첨 티켓 생성을 입력값을 기반으로 ManualLottoGenerator를 사용하도록 수정

programmer-sjk committed:
refactor: 출력 부분에서 getNumbers 대신 toString을 사용하도록 수정

객체의 메소드 대신 toString 사용
- 로또 티켓 번호 출력 로직에서 getNumbers 함수 제거
- toString 오버라이딩
- util 테스트 코드 추가

programmer-sjk committed:
refactor: 변수와 함수 네이밍 수정

mskangg commented on src/main/java/calculator/StringAddCalculator.java:
```diff
@@ -9,9 +9,11 @@
 public class StringAddCalculator {
```
피드백 반영 👍

mskangg commented on README.md:
```diff
@@ -25,4 +25,66 @@
\ No newline at end of file
+- [x] 음수를 전달할 경우 RuntimeException 예외가 발생해야 한다.
+
+## 3단계 - 로또(자동)
```
요구사항 및 기능 구현 목록 작성 부터 하시는것 좋습니다! 💯

mskangg commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.util.Scanner;
+
+public class InputView {
```
이 클래스는 static으로 이뤄져있는 유틸클래스로 보여집니다.
해당 클래스는 인스턴스화 되면 안되기에 
```java
private InputView() { }
```
로 생성자 접근을 막아야합니다.

이런 정적코드 분석 후 warning을 잡아주는 plugin 도 있습니다 :)
https://plugins.jetbrains.com/plugin/7973-sonarlint

programmer-sjk replied:
아하 감사합니다!

mskangg commented on src/main/java/lotto/view/ResultView.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.util.List;
+
+public class ResultView {
```
InputView 로 칭하였으니 OutputView 로 칭하는게 가독성 좋아보여요!

mskangg commented on src/main/java/lotto/util/RandomLottoGenerator.java:
```diff
@@ -0,0 +1,44 @@
+        List<Integer> randomNumbers = sort(toSubList(shuffle(numbers)));
+
+        return LottoTicket.create(randomNumbers.stream()
+                .map(i -> LottoNumber.get(i))
```
`.map(LottoNumber::get)` 람다로 변경가능합니다 :)

programmer-sjk replied:
핫..!! ㅋㅋㅋㅋ 감사합니다

mskangg commented on src/main/java/lotto/domain/enums/Rank.java:
```diff
@@ -0,0 +1,57 @@
+package lotto.domain.enums;
```
enums로 표현하시는 분도 계시는데, 개인적으로 enum도 싱글톤 객체라 다른 클래스와 동일하게 보고 있어 패키지를 나누진 않습니다! :)

mskangg commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -0,0 +1,49 @@
+
+import java.util.Objects;
+
+public class Money {
```
VO(값객체) 로 표현하신 부분인 것 같은데 맞을까요??
그렇다면, 객체 자체가 값으로 표현한다로 보여지는데, 값이 변경되게 구현하셨네요!

관련된 글 링크 남기겠습니다!
https://livenow14.tistory.com/18
https://tecoble.techcourse.co.kr/post/2020-06-11-value-object/

programmer-sjk replied:
아하,, 한 번 생성자에서 초기화 되면 값을 바꾸면 안되는군요. 감사합니다!!

mskangg commented on src/main/java/lotto/util/LottoGenerator.java:
```diff
@@ -0,0 +1,7 @@
+
+import lotto.domain.LottoTicket;
+
+public interface LottoGenerator {
```
LottoGenerator, ManualLottoGenerator, RandomLottoGenerator 이 3개를 util 패키지로 선언하신 이유가 있으신가요??
제가 생각하기엔 이 3개 모두 도메인 객체로 보여집니다!

programmer-sjk replied:
생성을 돕는다는 개념으로 util 패키지에 넣었는데, 도메인으로 이동하겠습니다. :)

mskangg commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,47 @@
+
+    private List<Integer> createWinningNumbers() {
+        String input = InputView.getWiningNumber()
+                .replaceAll(WHITE_SPACE_REGEX_STRING, EMPTY);
```
InputView 의 책임으로 보여집니다 :)

mskangg commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,47 @@
+        String input = InputView.getWiningNumber()
+                .replaceAll(WHITE_SPACE_REGEX_STRING, EMPTY);
+
+        return Arrays.asList(STANDARD_PATTERN.split(input)).stream()
```
`Arrays.stream(STANDARD_PATTERN.split(input))` 이처럼 List생성하고 stream 생성하지 않고 바로 호출할 수 있어요 :)

programmer-sjk replied:
감사합니다! 😄

mskangg commented on src/main/java/lotto/domain/dto/StatisticDto.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class StatisticDto {
```
`순위마다 몇 개 일치하는지 정보를 가짐` 이라 표현하셨는데, Dto 로 명시한 이유가 있으신가요?

당첨 결과를 갖고있는 객체를 만들어 상태로 갖고 있는 방향도 있어보입니다!

programmer-sjk replied:
LottoCommittee domain에 의해 생성되어 OutputView로 전달되고, 데이터 변환만 담당해서 dto를 붙였는데
LottoResult 객체를 추가하니 해당 클래스가 필요하지 않게 되었습니다. 이 방법이 좀 더 깔끔한 것 같아요!

mskangg commented on src/main/java/lotto/domain/LottoCommittee.java:
```diff
@@ -0,0 +1,30 @@
+
+    public StatisticDto statistics(List<LottoTicket> tickets) {
+        StatisticDto dto = StatisticDto.create();
+        for (LottoTicket ticket : tickets) {
+            dto.add(countOfMatch(ticket));
+        }
```
stream api 를 사용한다면
```java
tickets.forEach(lottoTicket -> dto.add(countOfMatch(lottoTicket)));
```
로 표현할 수 있겠네요!
이 부분은 취향으로 갈릴 것 같습니다

mskangg commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,41 @@
+    public static final int LOTTO_NUMBER_MIN = 1;
+    public static final int LOTTO_NUMBER_MAX = 45;
+
+    private static final List<LottoNumber> CACHES = IntStream.range(LOTTO_NUMBER_MIN, LOTTO_NUMBER_MAX)
```
미리 선언하고 꺼내 사용 💯

mskangg reviewed (changes requested):
안녕하세요 정국님!
새벽 5시까지,, 고생이 많으시네요!! 열정이 대단하십니다 👍 👍 👍 

우선 전체적으로 고민많이하신 티가 나는 코드로 보여집니다. 💯 
몇가지 어색한 부분이나 피드백이 필요한 부분 남겨놨으니 해당 부분 보완 부탁드립니다!

그럼 질문 사항 답변 남기겠습니다.

> Controller에 int money = ~~ 와 같은 코드가 나와 원시값 포장해야지. 하다가 모든 로직의 숫자를 Money 객체로 만들었는데, commit 후 돌아가기 힘든 강을 건넌 느낌입니다. 과연 좋은 방법일지 민수님은 어떻게 생각하실까요.

원시값 포장은 의미있는 행위입니다! 혹시 어떤 부분때문에 돌아가기 힘든 강을 건넌 느낌을 받으셨을까요??

---

> for 문이 Stream 보다 빠르니까 이해하기 쉬운 코드들은 for 문을 유지했는데 혹시 더 나이스하게 풀 수 있다면 꼭 리뷰를 남겨주시면 숙지 하겠습니다.

사실 이해하기 쉬운 코드는 현재 같이 일하는 팀원들이 `stream`을 주로 사용한다면, `stream` 메서드 체이닝 방식이 더 이해하기 쉬울 것이고, 반대라면 for 문이 이해하기 편할 것 입니다!
저희도 평소 단순 for문을 유지해서 개발 많이 합니다!
단, `filter, reduce, sum` 등 스트림에서 쉽게 제공하는 부분들을 사용할 수 있다면 stream을 차용합니다.

---

> 전체 상금을 계산하는 주체가, `Rank` 클래스가 상금 정보를 가지고 있어 `Rank` 클래스가 `calculatePrice` 메소드를 가지고 있는데 괜찮았을까요..?

당첨 결과를 관리하고 계산까지하는 클래스를 추가해서 해보시는건 어떠신가요?

예를 들어,

```java
private final Map<Rank, Integer> lottoResultCounts = new HashMap<>();
private final Lotto winningLotto;
```

를 상태로 갖고 있고 `calculatePrice` 를 구현하는 메서드도 갖고 있는 것이죠!

그렇다면 `Rank`와 `StatisticDto` 에 부여된 책임을 분리할 수 있어 보입니다.

programmer-sjk committed:
refactor: 리뷰 결과 반영

리뷰 사항 중 간단 개선건들 반영
- util -> domain 이동
- OutputView로 네이밍 변경
- Controller에 있던 replaceAll을 InputView로 이동
- Stream 문법 적용

programmer-sjk committed:
refactor: 리뷰 반영 (LottoResult domain)

통계 및 수익률을 담당하는 LottoResult 추가
- LottoCommittee 클래스 삭제
- StatisticDto 클래스 삭제
- LottoResult로 통계를 구하도록 로직 수정

programmer-sjk committed:
refactor: 리뷰 반영, 수익률 구하는 로직을 LottoResult에 위임

수익률 담당 책임을 LottoResult가 가지도록 수정
- LottoResult에 전체 가격을 계산하는 함수 추가
- Rank 클래스에서 계산하는 함수 제거
- 관련 테스트 코드 수정

programmer-sjk committed:
refactor: 리뷰 반영, 값 객체가 값을 변경하지 않도록 수정

programmer-sjk committed:
refactor: 함수 네이밍 수정

programmer-sjk commented:
`원시값 포장은 의미있는 행위입니다! 혹시 어떤 부분때문에 돌아가기 힘든 강을 건넌 느낌을 받으셨을까요??`
-> 아, 원시값 포장은 의미있지만, 전체 코드의 돈을  Money 객체로 포장하는게 옳은가?가령 Rank 클래스의 돈이 new Money()로 포장해도 좋은건가... 그런 고민이 있었습니다 ㅎㅎ; 그래서 Rank는 Money 객체가 아닌 숫자로 가지고 있을까? 돌려보려 했는데 제가 너무 많이 변경해서 그런 표현을 썼습니다.

피드백 중 `VO 객체의setter를 쓰지 않는다.`
-> 이 부분은 제가 몰랐던 부분인데 지적해주셔서 감사합니다.

programmer-sjk committed:
fix: map이 순서를 보장하도록 변경

mskangg reviewed (approved):
>  아, 원시값 포장은 의미있지만, 전체 코드의 돈을 Money 객체로 포장하는게 옳은가?가령 Rank 클래스의 돈이 new Money()로 포장해도 좋은건가... 그런 고민이 있었습니다 ㅎㅎ; 그래서 Rank는 Money 객체가 아닌 숫자로 가지고 있을까? 돌려보려 했는데 제가 너무 많이 변경해서 그런 표현을 썼습니다.

원시값을 포장하여 의미있는 값객체를 만들었다면, 해당 도메인 패키지 내에서 동작하는 비즈니스 로직들은 Money로 포장하고 사용해야한다고 생각해요.
다만, Rank에서의 Money는 구매를 위한 돈보단 단순 출력을 위한 상금으로 보여질 수 있기 때문에 Money가 안어울릴 수 있겠네요.
가령 구매가 1000원이 아닌 1 coin 으로 변경된다면 Money가 Coin이 될 것이고, 실제 상금은 돈으로 출력되어야 한다는 요구사항이 생기면 Money를 분리해야할 수도 있구요!

현재상황에선 어색하지 않게 잘하셨어요 👍 

> 이 부분은 제가 몰랐던 부분인데 지적해주셔서 감사합니다.

지적이라고 받아들이지 않으셨으면 좋겠습니다 😭😭😭
코드리뷰는 어디까지나 코드에 대한 생각을 공유하는 자리이며, 제가 하는 말들이 모두 정답이 아닌것처럼요!
서로 생각을 토론한다 생각해주시고 받아들여주셨으면 좋겠습니다! 🙏

피드백 반영 고생하셨습니다!
다음단계 진행해주세요~~ 🚀
